### ISSUE TITLE
[BUG] long fh does not work for specific forecasting pipeline

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#2807]
posted by @anthonygiorgio97 in #2452
(added some missing imports)
---

Another strange behaviour is that if `fh=[1,2]` the code below works, but if `fh=[1,2,3]` or `[1,2,3,4],...` or longer, I have error:

```python
from sktime.forecasting.compose import make_reduction, ForecastingPipeline, TransformedTargetForecaster
from sktime.forecasting.model_selection import ExpandingWindowSplitter, temporal_train_test_split, ForecastingGridSearchCV
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.transformations.series.difference import Differencer

from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVR
from sktime.datasets import load_longley

y, X = load_longley()

#split train/test both y and X
fh = [1,2,3]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

#pipeline
pipe = TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("differencer", Differencer(drop_na = True)),
            ("scaler", TabularToSeriesAdaptor(StandardScaler())),
            ("myforecasterpipe", ForecastingPipeline(steps = [
                ("logX", OptionalPassthrough(LogTransformer())),
                ("differencerX", Differencer(drop_na = True)),
                ("scalerX", TabularToSeriesAdaptor(StandardScaler())),
                ("myforecaster", make_reduction(SVR()))
        ]))  
])

#cv setup
N_cv_fold = 1
step_cv = 1
cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = fh)

param_grid = [{"log__passthrough": [False],
               "myforecasterpipe__logX__passthrough": [False],
               "myforecasterpipe__myforecaster__window_length": [2,3],
               "myforecasterpipe__myforecaster__estimator__C": [10,100]
              },
              {"log__passthrough": [True],
               "myforecasterpipe__logX__passthrough": [True],
               "myforecasterpipe__myforecaster__window_length": [2,3],
               "myforecasterpipe__myforecaster__estimator__C": [10,100]
              }
             ]

#grid search
gscv = ForecastingGridSearchCV(
            forecaster = pipe,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )

#fit
gscv.fit(train_model, X = X_train)
```

Error:

`ValueError: could not broadcast input array from shape (5,2) into shape (1,5,3)`

I think the problem regards the `inverse_transform` of `Differencer` on exogenous variable used for validation, when we have to do prediction in validation set. But I don't understand why for forecasting horizon `[1,2]` it works.

_Originally posted by @anthonygiorgio97 in https://github.com/alan-turing-institute/sktime/issues/2452#issuecomment-1100855032_

**Comment by fkiraly** [comment]
I can confirm that the bug still exists with 0.12.0

**Comment by fkiraly** [comment]
FYI @anthonygiorgio97, moved one of the unresolved bugs to a new issue.

**Comment by miraep8** [comment]
I will take a look at this bug!

**Comment by fkiraly** [comment]
Ah, interesting!

The original thread produced a similar bug:
https://github.com/alan-turing-institute/sktime/issues/2452
https://github.com/alan-turing-institute/sktime/pull/2487

In 0.13.0, the default of the `na_handling` param will change to `fill_zero`.

> That would help ensure that X ends up being the correct length at the end of the pipeline for the final prediction. 

I think the differencer is already doing that, the problem occurs in cases where there is no earlier time stamps to widen to. If you do backfilling, the result would be the same as `fill_zero`.

Having read the above, I'm not sure what the best action is. `fill_zero` default prevents the condition from materializing, but if the user chooses to shorten the `X`, the error message is not informative.

Questions one might ask:
* is an action necessary at all? E.g., should a user know that the `Differencer` shortens series?
* The `Differencer` docstring could be improved with a warning that certain parameter settings shorten the series, others create NA (and pipelining with an `Imputer` may be advised)
* if a generic error message about index incompatibility should be raised, where should it be raised? Could be a warning along the lines of "transformer A does not guarantee same indices as output, but forecaster B requires same index set as input"?

**Comment by miraep8** [comment]
> I think the differencer is already doing that, the problem occurs in cases where there is no earlier time stamps to widen to. If you do backfilling, the result would be the same as fill_zero.

I guess it might be worth verifying that this works properly in the context of a pipeline or something as well?  For example - if `Differencer` is the third step in a transformer pipeline then we would want to grab the extra points for backfilling before the first step so they get properly transformed.

>if a generic error message about index incompatibility should be raised, where should it be raised? Could be a warning along the lines of "transformer A does not guarantee same indices as output, but forecaster B requires same index set as input"?

Actually broader question related to this - is there a reason `sktime` doesn't make broader use of the `logger` class? This seems like it *could* be an appropriate time to log a warning, and then if there is some sort of error, I *think* the relevant warning (and tips) would print out with the error trace?

**Comment by fkiraly** [comment]
> For example - if `Differencer` is the third step in a transformer pipeline then we would want to grab the extra points for backfilling before the first step so they get properly transformed.

Good point. Though, I can't see a way for this without "contextual awareness" of estimator, i.e., the `Differencer` being aware that it's being called within a pipeline and what the other steps are. Contextual awareness of estimators is imo one of many pathways to hell, can you think of a way that would not require it?

>Actually broader question related to this - is there a reason `sktime` doesn't make broader use of the `logging` class? 

I think it's a combination of laziness and lack of developer familiarity (e.g., I would think I have only beginner familiarity with `logging`).

With the "plumbing" layer in place, we could localize logging concerns there? Any good ideas, e.g., how a design might look like, and how it would materialize in this specific example?

Re this very bug: do you think it's solved by the changes to `Differencer` parameters, or is more action needed?

**Comment by miraep8** [comment]
>Good point. Though, I can't see a way for this without "contextual awareness" of estimator, i.e., the Differencer being aware that it's being called within a pipeline and what the other steps are. Contextual awareness of estimators is imo one of many pathways to hell, can you think of a way that would not require it?

I totally agree! I think my plan was not to make `Differencer` contextually away but pipelines!  ie - pipelines are aware of all the steps and *in theory* should be able to count the number of lost columns and adjust the input accordingly.  To me this seems like it could be an appropriate place to do so, then we could encourage anyone chaining together transformers/forecasters to use the pipelining features.  Or we go the `pywatts` route which is closer to everything is a pipeline? (even if just one step?)

> Re this very bug: do you think it's solved by the changes to Differencer parameters, or is more action needed?

I think it is "solved" by no longer raising errors, but I suspect statistically it might not be doing what the original author intended... (due to the issues above) so I would potentially be hesitant to mark it as fully solved.  I have a potential workaround in mind that might not require changing the code base (but is in-elegant) by just splitting the pipeline into 2 parts - a transformer-style pipeline that reduces the dimension by 1 (ie starts out with fh one larger) and then feeding the output of that pipeline into a `TransformedTargetForecaster` which has the remainder of the original steps. Would need to check that it actually works before posting it as a solution. 

As to logging, I think Lukasz has an example of what that might look like in sktime in his PR #2744. For this example I was thinking about logging a warning when the user uses the `drop_na` option that it might raise issues.  The idea would be to create a more informative error trace if there are issues down the road from making that choice

**Comment by fkiraly** [comment]
>pipelines are aware of all the steps and in theory should be able to count the number of lost columns and adjust the input accordingly.

If that's the plan, then the components need to be able to tell the pipeline what the expected output *size* is, not just the format. E.g., "which columns will you lose"?
I was already thinking about giving transformers a method such as `get_expected_cols(X)` or similar, and one could have `get_expected_rows`, although I wonder whether it really adds sth above simply passing `X` to `transform` and looking at what comes out.

>I have a potential workaround in mind that might not require changing the code base (but is in-elegant) by just splitting the pipeline into 2 parts - a transformer-style pipeline that reduces the dimension by 1 (ie starts out with fh one larger) and then feeding the output of that pipeline into a TransformedTargetForecaster which has the remainder of the original steps. Would need to check that it actually works before posting it as a solution.

I don't understand the proposed solution here. You would have some transformers separate? Based on what would we split the transformer chain? Can you explain? Perhaps also in one of the meetings, since it seems involved.

>As to logging, I think Lukasz has an example of what that might look like in sktime in his PR https://github.com/alan-turing-institute/sktime/pull/2744. For this example I was thinking about logging a warning when the user uses the drop_na option that it might raise issues. The idea would be to create a more informative error trace if there are issues down the road from making that choice

Maybe it's also worth looking at the mock estimator(s) of @ltsaprounis which already log method calls?

**Comment by miraep8** [comment]
So, I wanted to follow up on this issue, both to propose a solution to this specific problem, and to move the design questions which have come up as a result of this issue to a new more design-focused issue (#2880). 

First, here is a proposed solution which works with our current code base.  Essentially one can pre-transform the data for all the steps up to and including the one which changes size.  Then you can continue with the rest of the pipeline as before.  One disadvantage of this solution is that it makes it harder to automate either including or not including the `Log-Transformer` with `OptionalPassthrough`  In the meantime I suggest the original user who raised this issue compare the best solutions available both with and without that transformer manually.  (and hopefully we will arrive at a cleaner solution soon)

```
from sktime.forecasting.compose import make_reduction, ForecastingPipeline, TransformedTargetForecaster
from sktime.forecasting.model_selection import ExpandingWindowSplitter, temporal_train_test_split, ForecastingGridSearchCV
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.transformations.series.difference import Differencer
from sktime.transformations.compose import TransformerPipeline
from sklearn import clone

from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVR
from sktime.datasets import load_longley

y, X = load_longley()

#split train/test both y and X
fh = [1,2,3,4]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

#preprocess the steps
pre_process_steps = [
    LogTransformer(),
    Differencer(),
]

y_transformed = train_model
X_transformed = X_train
for transformer in pre_process_steps:
    y_transformed = clone(transformer).fit_transform(X=y_transformed)
    X_transformed = clone(transformer).fit_transform(X=X_transformed)

#pipeline
pipe = TransformedTargetForecaster(steps = [
            ("scaler", TabularToSeriesAdaptor(StandardScaler())),
            ("myforecasterpipe", ForecastingPipeline(steps = [
                ("scalerX", TabularToSeriesAdaptor(StandardScaler())),
                ("myforecaster", make_reduction(SVR()))
        ]))  
])

#cv setup
N_cv_fold = 1
step_cv = 1
cv = ExpandingWindowSplitter(initial_window = len(y_transformed) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = fh)

param_grid = [{"myforecasterpipe__myforecaster__window_length": [2,3],
               "myforecasterpipe__myforecaster__estimator__C": [10,100]
              },
              {"myforecasterpipe__myforecaster__window_length": [2,3],
               "myforecasterpipe__myforecaster__estimator__C": [10,100]
              }
             ]

#grid search
gscv = ForecastingGridSearchCV(
            forecaster = pipe,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )

#fit
gscv.fit(y_transformed, X = X_transformed)
```
As to more design minded questions - we are moving that discussion to a new issue (#2880), however in the meantime I suspect one reasonable solution may be that even though we are using these pipelines as forecasters, we might not actually need to trim X for the actual forecasting step until the final step which actually performs the forecasting?

**Comment by miraep8** [comment]
@fkiraly, if you agree that the above solution is reasonable, maybe we can close this issue?

**Comment by fkiraly** [comment]
> we might not actually need to trim X for the actual forecasting step until the final step which actually performs the forecasting?

Hm, not sure - I thought the issue is not so much with `X` being *trimmed* but being *transformed* into sth with a different index set. Where there is not a clear "not trimmed" option, since the result is not simply `X` at some sub-set of columns.

**Comment by miraep8** [comment]
Yes!  Sorry, I think my initial message was unclear!

I wanted to just loop in the original poster of this bug @anthonygiorgio97 and bring you up to speed!

First off, wanted to thank you for bringing up this bug!  It has brought up some interesting design questions (as you can see).  We are having on ongoing design discussion as to how to address this problem in a more permanent way (#2880 if you would be interested in joining in!) but in the meantime have found a temporary workaround solution that might work for you! (it is located in the comment 3 above this one!).  Main takeaways and things to be aware of:
- This work around doesn't allow you to do full tuning of any transformers not in `TransformedTargetForecaster`.  In order to try out both including and not including `LogTransformer` like you were originally trying to do I would recommend comparing the performance of the best tuned models while either applying that transformer (as in the text above) or commenting it out. 
- A better solution is in the works - though it might be a while before it gets fully resolved.  We intend to update this issue when the matter gets fully resolved. 

Please let us know if you have any further questions, ideas or complaints about this issue in the meantime though! :)

**Comment by benHeid** [comment]
Neither the original code posted by @fkiraly (after small fixes) is reproducing the issue nor the code provided by Claude Code. Since the related design decisions are part of issue #2880. I am going to close this issue.


<details>

```python

import numpy as np
import pandas as pd

from sktime.forecasting.compose import (
    make_reduction,
    ForecastingPipeline,
    TransformedTargetForecaster,
)
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.transformations.series.difference import Differencer

from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVR

rng = np.random.default_rng(42)
n = 20
index = pd.period_range("2000", periods=n, freq="Y")
y = pd.Series(np.linspace(100, 120, n) + rng.normal(0, 0.5, n), index=index, name="y")
X = pd.DataFrame(
    {
        "x1": np.linspace(50, 70, n) + rng.normal(0, 0.5, n),
        "x2": np.linspace(10, 30, n) + rng.normal(0, 0.5, n),
    },
    index=index,
)

fh = [1, 2, 3]
y_train, y_test = y.iloc[: -len(fh)], y.iloc[-len(fh):]
X_train, X_test = X.iloc[: -len(fh)], X.iloc[-len(fh):]

pipe = TransformedTargetForecaster(
    steps=[
        ("log", OptionalPassthrough(LogTransformer())),
        ("differencer", Differencer(na_handling="drop_na")),
        ("scaler", TabularToSeriesAdaptor(StandardScaler())),
        (
            "myforecasterpipe",
            ForecastingPipeline(
                steps=[
                    ("logX", OptionalPassthrough(LogTransformer())),
                    ("differencerX", Differencer(na_handling="drop_na")),
                    ("scalerX", TabularToSeriesAdaptor(StandardScaler())),
                    ("myforecaster", make_reduction(SVR(), window_length=2)),
                ]
            ),
        ),
    ]
)
pipe.set_params(log__passthrough=False, myforecasterpipe__logX__passthrough=False)

print(
    "Fitting TransformedTargetForecaster + ForecastingPipeline with "
    f"Differencer(na_handling='drop_na') on synthetic data (n={n}), fh={fh}"
)

caught_error = None
y_pred = None
try:
    pipe.fit(y_train, X=X_train)
    y_pred = pipe.predict(fh=fh, X=X_test)
    print("Fit and predict succeeded. Prediction:")
    print(y_pred)
except ValueError as e:
    caught_error = e
    print(f"ValueError raised during fit/predict: {e}")

assert caught_error is None, (
    "Reported bug reproduced: fitting/predicting a TransformedTargetForecaster containing a "
    "ForecastingPipeline with a size-changing Differencer(na_handling='drop_na') on exogenous X "
    f"raised a ValueError for fh={fh} (length 3): {caught_error}"
)
assert y_pred is not None and len(y_pred) == len(fh), (
    f"Expected {len(fh)} predicted values, got {0 if y_pred is None else len(y_pred)}"
)
print("No error raised and forecast horizon length matches fh - bug appears fixed.")
```

</details>

Triaged with the help of Claude Code
